### PR TITLE
[Load/Save] Update to load and save the adam momentum variables

### PR DESCRIPTION
**Pull request by jijoongmoon**
In this PR, Load and save is updated for the adam optimizer variables.
In order to do it, optimizer load twice, one for the weight, and one
for the optimizer variables.

**Changes proposed in this PR:**
- Added TOC generator for README.md

Resolves:

**Self evaluation:**
1. Build test:	 [X]Passed [ ]Failed [ ]Skipped
2. Run test:	 [X]Passed [ ]Failed [ ]Skipped

### CODE REVIEW

File: nntrainer/models/neuralnet.cpp [hunk 1]
```diff
@@ -205,7 +205,10 @@ int NeuralNetwork::initialize() {
 
   initialized = true;
 
-  if (!load_path.empty()) {
+  // @note we need check loadedWeight for the case of multiple call of load to
+  // load weight. Only the weight needs to be loaded here. Becuase the buffer
+  // for the optimizer is not allocated yet.
+  if (!load_path.empty() && !loadedWeight) {
```
Comment (kparichay) [inline nntrainer/models/neuralnet.cpp:211]: is this possible? in the current design, the model cannot load weights before initialization as initialization allocates weight buffers.

File: nntrainer/layers/layer_node.cpp [hunk 2]
```diff
@@ -420,10 +420,13 @@ void LayerNode::read(std::ifstream &file, bool opt_var) {
     << __func__ << " layer needs to be finalized first!";
   if (opt_var) {
     for (unsigned int i = 0; i < run_context->getNumWeights(); ++i) {
-      if (run_context->isGradientLastAccess(i)) {
+      if (run_context->isGradientLastAccess(i) && getTrainable()) {
         // @note read optimizer variables
-        for (unsigned int j = 0; j < run_context->getNumWeightOptVar(i); ++j) {
-          run_context->getWeightOptVar(i, j).read(file);
+        if (run_context->weightHasGradient(i)) {
```
Comment (kparichay) [inline nntrainer/layers/layer_node.cpp:425]: weightHasGradient() and getTrainable() have the same effect?

Comment (jijoongmoon) [reply nntrainer/layers/layer_node.cpp:425]: it is not. for the batch normalization layer, it has 4 weights and two of them do not have gradient even if trainable is set true.

File: nntrainer/models/neuralnet.cpp (hunk 1 above)
Comment (jijoongmoon) [reply nntrainer/models/neuralnet.cpp:211]: model_graph.allocateWeights() is called at line 204. so, the buffer for weight is allocated and so, it is possible to load weight.

File: nntrainer/layers/layer_node.cpp (hunk 2 above)
Comment (kparichay) [reply nntrainer/layers/layer_node.cpp:425]: thanks for the explanation.

File: nntrainer/models/neuralnet.cpp (hunk 1 above)
Comment (kparichay) [reply nntrainer/models/neuralnet.cpp:211]: My concern is `loadedWeight` will always be false here, as the model cannot be initialized twice and NeuralNetwork::load() needs model to be initialized.

**Review by kparichay — APPROVED** [review]
LGTM

File: nntrainer/models/neuralnet.cpp [hunk 3]
```diff
@@ -381,18 +385,22 @@ void NeuralNetwork::load(const std::string &file_path,
...
       model_file.seekg(bin_file_pos);
 
-      std::string opt_type;
-      opt_type = readString(model_file);
-      if (istrequal(opt_type, "adam") && istrequal(opt->getType(), "adam")) {
-        for (auto iter = model_graph.cbegin(); iter != model_graph.cend();
-             iter++) {
-          (*iter)->read(model_file, true);
+      if (istrequal(opt->getType(), "adam")) {
+        char opt_type[4];
```
Comment (lhs8928) [inline nntrainer/models/neuralnet.cpp:397]: Does the adam is the only case which loads variables for optimizer?
If it is not, then I think the previous way is more suitable for general cases.

**Review by lhs8928 — APPROVED** [review]
LGTM except minor comment

**Pull request merged**